cpp/src/binaryop/compiled/operation.cuh

```diff
@@ -0,0 +1,384 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/types.hpp>
+#include "traits.hpp"
+
+#include <cmath>
+
+#include <cuda/std/type_traits>
+
+using namespace cuda::std;
+
+namespace cudf {
+namespace binops {
+namespace compiled {
+
+/**
+ * @brief Type casts each element of the column to `CastType`
+ *
+ */
+template <typename CastType>
+struct type_casted_accessor {
+  template <typename Element>
+  CUDA_DEVICE_CALLABLE CastType operator()(cudf::size_type i, column_device_view const col) const
+  {
+    if constexpr (column_device_view::has_element_accessor<Element>() and
+                  std::is_convertible_v<Element, CastType>)
+      return static_cast<CastType>(col.element<Element>(i));
+    return {};
+  }
+};
+
+/**
+ * @brief Type casts value to column type and stores in `i`th row of the column
+ *
+ */
+template <typename FromType>
+struct typed_casted_writer {
+  template <typename Element>
+  CUDA_DEVICE_CALLABLE void operator()(cudf::size_type i,
+                                       mutable_column_device_view const col,
+                                       FromType val) const
+  {
+    if constexpr (mutable_column_device_view::has_element_accessor<Element>() and
+                  std::is_constructible_v<Element, FromType>) {
+      col.element<Element>(i) = static_cast<Element>(val);
+    } else if constexpr (is_fixed_point<Element>() and is_fixed_point<FromType>()) {
+      col.data<Element::rep>()[i] = val.value();
+    }
+  }
+};
+
+// All binary operations
+namespace ops {
+
+struct Add {
```

Comment:
It's not used in host in our code. These ops are valid for host calls too. It's available for anyone to used it in host if needed. So it is kept as both device and host callable. Since they are templates, they don't take symbol space in binary unless instantiated in host.
I will check if device side callable only helps compilation time.

Reply:
I see you moved all the ops kernels over to `CUDA_DEVICE_CALLABLE` did it improve compile performance?

Reply:
No noticeable change.